Code/GraphMol/Descriptors/BagOfBonds.cpp

```diff
@@ -0,0 +1,310 @@
+      std::advance(it1,i);
+      std::pair<std::string, unsigned int> Values1 = *it1;
+      for (unsigned int j = i+1; j < nbElements ; j++) {
+        std::map<std::string, unsigned int>::const_iterator it2= counts.begin();
+        std::advance(it2,j);
+        std::pair<std::string, unsigned int> Values2 = *it2;
+        std::string PairString;
+        // sorting string order
+        if (Values1.first > Values2.first) {
+           PairString = Values2.first+Values1.first;
+        }
+        else {
+            PairString = Values1.first+Values2.first;
+        }
+        unsigned int nbElementPair = Values1.second*Values2.second;
+        std::pair<std::string,unsigned int > key = std::make_pair(PairString, nbElementPair);
+        ks.push_back(key);
+      }
+  }
+
+  for (std::map<std::string, unsigned int>::const_iterator countIt = counts.begin();
+       countIt != counts.end(); ++countIt) {
+    //"homologues":
+    unsigned int nbElementAtom = countIt->second;
+    unsigned int res =nbElementAtom*(nbElementAtom-1)/2;
+    std::string PairString = countIt->first+countIt->first;
+    if (res>0) {
+      std::pair<std::string,unsigned int > key = std::make_pair(PairString, res);
+      ks.push_back(key);
+    }
+  }
+  return ks;
+
+}
+
+
+ BagMatrix getCoulombMatBags(Eigen::VectorXd numbers, int numatoms, Eigen::MatrixXd Distance3D,  std::vector<std::string > S, int alpha) {
+  // 3D distance matrix
+  Eigen::MatrixXd ProdV = numbers*numbers.transpose(); // outer products of vector ie (z[i] * z[j])
+
+  if (alpha != 1) {
+    Distance3D = Distance3D.array().pow(alpha); // power distance alpha
+  }
+
+  Eigen::MatrixXd MyVal = ProdV.cwiseQuotient(Distance3D); // ratio top / dist
+  MyVal.diagonal() = 0.5 * numbers.array().pow(2.4); // set the diagonal fix values 
+  // get the BagsCode
+  std::vector<std::string > BagsCodes;
+  std::list<std::pair<std::string, double > > Res;
+  std::string s1, s2, key;
+  for (unsigned int i = 0; i< numatoms; i++) {
+    s1 = S[i];
+    BagsCodes.push_back(s1); // adding the diagonal once!
+
+    for (unsigned int j = i+1; j< numatoms; j++ ){
+      s2 = S[j];
+      if (s1 > s2) {
+        key = s2+s1;
+      }
+      else {
+        key = s1+s2;
+      }
+      BagsCodes.push_back(key);
+    }
+  }    
+
+  Eigen::VectorXd RES = getUpperMat(MyVal);
+  return BagMatrix{RES , BagsCodes}; 
+}
+
+///// code to be expose to python directly not simple
+// we can parallelized cause we need the max value per key
+ std::map<std::string, unsigned int> getBoBSizefromSmiles(std::vector<std::string> smiles) {
+    //unsigned int idx = 0;
+    //std::vector< std::list < std::pair<std::string, unsigned int> > > myData(smiles.size());
+    std::list<std::pair<std::string, unsigned int> > fulldata;
+    std::map<std::string, unsigned int> Global;
+    std::string key;
+
+    //while (idx < smiles.size()) {
+    for (auto const& smi :  smiles) {
+      RDKit::ROMol *mol;
```

Comment:
might as well use an `std::unique_ptr` here, then you don't need to worry about lifetime